From c763f983cc1532b4ca206ba46ae8150493dcbf90 Mon Sep 17 00:00:00 2001
From: Weiwei Wang <weiwei.wang@windriver.com>
Date: Fri, 1 Jul 2011 17:21:25 +0800
Subject: [PATCH 1/2] net/dpa: replace local_irq* with local_irq*_nort

In preempt_rt kernel, sometimes following call trace appears:
BUG: sleeping function called from invalid context at kernel/rtmutex.c:707
pcnt: 0 0 in_atomic(): 0, irqs_disabled(): 1, pid: 924, name: modprobe
Call Trace:
[e9fa9ba0] [c0007c20] show_stack+0x44/0x160 (unreliable)
[e9fa9bd0] [c00350a8] __might_sleep+0xe4/0x108
[e9fa9be0] [c05f5dd4] rt_spin_lock+0x38/0xb4
[e9fa9bf0] [c010022c] _slab_irq_disable+0x50/0x8c
[e9fa9c20] [c01023c8] __kmalloc+0x80/0x268
[e9fa9c60] [c047464c] xx_Malloc+0x20/0x78
[e9fa9c70] [c040ce90] DtsecAddHashMacAddress+0x11c/0x2b4
[e9fa9ca0] [c041360c] FM_MAC_AddHashMacAddr+0x2c/0x140
[e9fa9cb0] [c05f6c00] set_multi+0x124/0x2b4
[e9fa9d00] [c04756a4] dpa_set_multicast_list+0x30/0x100
[e9fa9d20] [c05437d0] __dev_set_rx_mode+0x44/0xb4
[e9fa9d40] [c054cfd4] dev_mc_add+0x80/0xa0
[e9fa9d60] [f96962b4] igmp6_group_added+0xf4/0x1a8 [ipv6]
[e9fa9da0] [f96966f0] ipv6_dev_mc_inc+0x328/0x4cc [ipv6]
[e9fa9dd0] [f967aff0] ipv6_add_dev+0x22c/0x344 [ipv6]
[e9fa9df0] [f967ce4c] addrconf_notify+0x64/0x864 [ipv6]
[e9fa9e90] [c0547898] register_netdevice_notifier+0xb4/0x204
[e9fa9ec0] [f959233c] addrconf_init+0x80/0x160 [ipv6]
[e9fa9ee0] [f959217c] inet6_init+0x17c/0x294 [ipv6]
[e9fa9ef0] [c0001d48] do_one_initcall+0x3c/0x1e0
[e9fa9f20] [c0086970] sys_init_module+0xf8/0x21c
[e9fa9f40] [c00114b4] ret_from_syscall+0x0/0x4
Exception: c00 at 0xff63564
LR = 0x100031b8

the root cause is xx_Malloc is called with irq disabled. To fix it,
local_irq*_nort should replace local_irq*.

Signed-off-by: Weiwei Wang <weiwei.wang@windriver.com>
---
 drivers/net/dpa/NetCommSw/src/xx/xx_linux.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/net/dpa/NetCommSw/src/xx/xx_linux.c b/drivers/net/dpa/NetCommSw/src/xx/xx_linux.c
index 5f51336..f4cbb9b 100644
--- a/drivers/net/dpa/NetCommSw/src/xx/xx_linux.c
+++ b/drivers/net/dpa/NetCommSw/src/xx/xx_linux.c
@@ -368,14 +368,14 @@ uint32_t XX_DisableAllIntr(void)
 {
     unsigned long flags;
 
-    local_irq_save(flags);
+    local_irq_save_nort(flags);
 
     return (uint32_t)flags;
 }
 
 void XX_RestoreAllIntr(uint32_t flags)
 {
-    local_irq_restore((unsigned long)flags);
+    local_irq_restore_nort((unsigned long)flags);
 }
 
 t_Error XX_Call( uint32_t qid, t_Error (* f)(t_Handle), t_Handle id, t_Handle appId, uint16_t flags )
-- 
1.7.0.4

